plugins/reindex/src/main/java/org/elasticsearch/plugin/reindex/ReindexRequest.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.plugin.reindex;
+
+import static org.elasticsearch.action.ValidateActions.addValidationError;
+import static org.elasticsearch.index.VersionType.INTERNAL;
+
+import java.io.IOException;
+
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.index.IndexRequest;
+import org.elasticsearch.action.search.SearchRequest;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.lucene.uid.Versions;
+
+public class ReindexRequest extends AbstractBulkIndexByScrollRequest<ReindexRequest> {
+    /**
+     * Prototype for index requests.
+     */
+    private IndexRequest destination;
+
+    public ReindexRequest() {
+    }
+
+    public ReindexRequest(SearchRequest search, IndexRequest destination) {
+        super(search);
+        this.destination = destination;
+    }
+
+    @Override
+    protected ReindexRequest self() {
+        return this;
+    }
+
+    @Override
+    public ActionRequestValidationException validate() {
+        ActionRequestValidationException e = super.validate();
+        if (source().indices() == null || source().indices().length == 0) {
+            e = addValidationError("use _all if you really want to copy from all existing indexes", e);
+        }
+        /*
+         * Note that we don't call index's validator - it won't work because
+         * we'll be filling in portions of it as we receive the docs. But we can
+         * validate some things so we do that below.
+         */
+        if (destination.index() == null) {
+            e = addValidationError("index must be specified", e);
+            return e;
+        }
+        if (false == (destination.routing() == null || destination.routing().startsWith("=") ||
```

Comment:
I think this check is a bit hard to read. Maybe extract methods and/or reformat?